### PR TITLE
Converter block storage

### LINKED ISSUE #1505: Calling update after write_to fails
The following example fails in asdf 2.14.4 (and the current main head)
```python
import asdf
import numpy

fn1 = 'test1.asdf'
fn2 = 'test2.asdf'

tree = {'a': numpy.zeros(3), 'b': numpy.ones(3)}
af = asdf.AsdfFile(tree)

af.write_to(fn1)

with asdf.open(fn1, mode='rw') as af:
    af['a'] = None
    af.write_to(fn2)
    af.update()
```
with the following error:
```
Traceback (most recent call last):
  File "/Users/bgraham/projects/230130_asdf_2p15_prep/tests/write_to_then_update/test.py", line 15, in <module>
    af.update()
  File "/Users/bgraham/.pyenv/versions/asdf_2p15_prep/lib/python3.10/site-packages/asdf/asdf.py", line 1225, in update
    self._blocks.finish_reading_internal_blocks()
  File "/Users/bgraham/.pyenv/versions/asdf_2p15_prep/lib/python3.10/site-packages/asdf/block.py", line 286, in finish_reading_internal_blocks
    last_block = self._read_next_internal_block(last_block._fd, False)
  File "/Users/bgraham/.pyenv/versions/asdf_2p15_prep/lib/python3.10/site-packages/asdf/block.py", line 227, in _read_next_internal_block
    block = self._new_block().read(fd, past_magic=past_magic, validate_checksum=self._validate_checksums)
  File "/Users/bgraham/.pyenv/versions/asdf_2p15_prep/lib/python3.10/site-packages/asdf/block.py", line 1074, in read
    raise ValueError(msg)
ValueError: Bad magic number in block. This may indicate an internal inconsistency about the sizes of the blocks in the file.
```
Removing the call to `write_to` or `update` succeeds.

This was found during working on tests for a "new block manager on write" which is likely one solution to this issue.

### PR DESCRIPTION
**Pull request by braingram**
This PR adds block storage support to Converters.

Relevant docs are here: https://asdf--1508.org.readthedocs.build/en/1508/asdf/extending/converters.html#block-storage

The key contribution of this PR is allowing Converters to read/write blocks during from_yaml_tree/to_yaml_tree. This is accomplished by (privately) providing the BlockManger to the SerializationContext (that is passed as an argument to Converter methods) and introducing a few new public methods:
- Converter.reserve_blocks
- ~SerializationContext.load_block~ EDIT: removed in 92a786a9045f16c490907af05e09a199d6440570
- SerializationContext.get_block_data_callback EDIT: added in 92a786a9045f16c490907af05e09a199d6440570
- SerializationContext.assign_block_key
- SerializationContext.find_block_index
- asdf.util.BlockKey EDIT: added in a7193a11536bc6de0040b5ec54cedf4f30c7355a

The docs update describes the public API and several tests were added to cover these new methods and features (this is also mostly compatible with the [asdf_zarr chunking prototype](https://github.com/braingram/asdf_zarr/blob/deferred_block/src/asdf_zarr/converter.py) with one exception, the storage settings, described below).

There was one major change required to get this to work. With this PR, when `AsdfFile.write_to` is called, a new `AsdfFile` object is created to be used during write. This allows write to modify blocks (in this case overwriting data callbacks used to defer reading data until a block is ready to be written) without effecting blocks created during initial reads of a file. This change required updates to many tests which expected the blocks to change after `write_to`.

A few minor changes include:
- Fixes #1505 (and adds regression test `test_write_to_before_update`)
- Don't limit IntegerType _value_cache per ctx/AsdfFile
- Ignore `ERA001` rule that is supposed to remove commented code, this failed several times removing some but not all of the commented code so I just disabled it

There are a few things missing from or related to this PR that should be discussed and likely included as follow-up PRS.
- storage settings. I initially had these included and modeled after the existing array settings (set_block_storage == set_array_storage, etc). I removed them because I think a more flexible and simpler interface will be preferred. The [asdf_zarr prototype used/uses these settings](https://github.com/braingram/asdf_zarr/blob/f9c850959db4834bf8586f5f88123e5456d37b01/src/asdf_zarr/converter.py#L124) to determine if a zarray should be included as internal blocks or left as an external zarray. However this required adding additional settings to the block converter, functions to the AsdfFile and to the SerializationContext to cover all places where these settings are used. This did not seem ideal and did not allow for settings that didn't match those used for arrays (setting storage to 'internal', 'external', 'inline', etc). I have a vague idea that this might be a useful way to provide additional options to configure the converter and discussing and deciding on an API seems worthwhile (even if it ends up being a copy of what's used for arrays).
- untangle uses of converter methods and reuse of blocks. Currently (and with this PR) there are a lot of multi-use objects within the ASDF internal. These made the updates in this PR  difficult and often brittle as a small change to Block (for example) had the possibility of changing reading, writing, compression etc. I spent some time trying to break apart some of the main offenders (mainly the block management) but none of that resulted in a simpler PR. Here are a few other examples:
  - from_yaml_tree is used for reading, fil_defaults, remove_defaults and is unaware of the context making it difficult to (for example) know if blocks should be created
  - to_yaml_tree is used for writing, validate, fill_default, remove_defaults, twice during update (once to judge tree size)
  - blocks are used for storing settings (compression storage), sometimes data (sometimes a file descriptor), file state (offset, allocated size), and for both reading and writitng
I will open issues for the above if reviews do not lead to related changes or fixes.

### CODE REVIEW

**Comment by braingram** [comment]
Based off the discussion during the ASDF tag-up I replaced `load_block` which required holding onto the `SerializationContext` with `generate_block_data_callback` to allow lazy loading without holding onto a reference to the context. And the use of `id(obj)` in the examples and tests to avoid issues with memory reuse.

Thanks @eslavich and @perrygreenfield

**Comment by braingram** [comment]
Apologies in advance for what will probably be a novel.

TLDR; the [asdf_zarr prototype](https://github.com/braingram/asdf_zarr/tree/deferred_block) is now up to date with this PR and is ready to be moved to asdf-format. Also, a follow up PR might allow us to remove the need for reserved blocks.

I've spent a few days trying to sort out a reasonable set of functions to expose the storage options to converters. This would include changes to `AsdfFile` (where settings are often set by the user), `SerializationContext` (which is available to the converter) and probably `BlockManager` as this serves storage of settings (at the moment). The settings are complicated by the key/id issue mentioned above. There are a few points I'd like to bring up after looking into this.

First, the way I was using the settings previously in the asdf_zarr prototype is likely not something we should aim to support. I was looking at the array storage to determine if the zarr chunk store should be included as internal asdf blocks (if the array storage was 'internal') or left as an external store (like a DirectoryStore, if the array storage was 'external'). This conflates options for the Converter and options for the blocks. I rewrote the asdf_zarr to instead require a conversion of zarr-arrays with a [to_internal](https://github.com/braingram/asdf_zarr/blob/3a79f1cc3241a7ff9597eba013703cad746eb156/src/asdf_zarr/storage.py#L60) function (without copying data until the asdf file is written). There might be a wider discussion of if we want to provide a means of providing options to converters but I think that this should not be required for chunked array support.

Second, the block management is complicated (as mentioned above) and the `BlockKey` introduced in this PR seems workable for our current needs. However, I think the following alternative strategy might be better but would require substantial rework of the block management. Currently the block keys are required because memory addresses might be reused such as with the following:
```python
af = asdf.open(fn, mode="rw")  # open a file with a custom object 'a' at id(obj) == 42
del af['a']  # this does not delete the blocks that would be associated with 'a'
b = Foo()  # make a new object, let's assume python puts it at id(b) == 42
af['b'] = b  # now the blocks for 'a' will be incorrectly assigned to 'b'
```
The alternative strategy would be to have asdf internally manage keys for each block that hold weakrefs to objects (here is a [partial example](https://github.com/braingram/asdf/blob/storage_options/asdf/util.py#L535) from some of my exploration). This would allow objects to be garbage collected when they are removed from the tree and be resilient to memory address reuse (as 'matching' the key would involve resolving the weakref which would return None for the missing object). This has the benefit of not requiring that developers of converters generate their own BlockKey instances and since Converter.from_yaml_tree returns the object that I think we would want to attach to the blocks we might be able to have ASDF check which blocks were accessed during from_yaml_tree, take the returned object and create and assign block keys outside the converter. This should mean that `reserve_blocks` is no longer needed and possibly `assign_block_key` as well (we might still want some way of assigning keys within the converter because, as with the asdf_zarr prototype, a single object might use multiple blocks which may need to be retrieved during to_yaml_tree for an update). I've made a few attempts at getting this to work without rewriting the BlockManager with no success.

File: asdf/asdf.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def __init__(
         self._closed = False
         self._external_asdf_by_uri = {}
         self._blocks = block.BlockManager(self, copy_arrays=copy_arrays, lazy_load=lazy_load)
-        self._uri = None
+        self._uri = uri
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/asdf.py:152]: Thanks! We do not need the second bit of code. The change here was made while developing the new AsdfFile on write. During development, I incorrectly provided the tree to AsdfFile.__init__ (this results in an unnecessary validation) which revealed that passing a tree AsdfFile.__init__ when all_array_storage is set to external due to this assignment triggering a validation (prior to the uri being assigned):
https://github.com/asdf-format/asdf/blob/9bcdbbefe80ed1c0bcc8fd136db49381d1c0b4f4/asdf/asdf.py#L169
I just added a regression test for this fix, a note about the assignment and removed the unnecessary second assignment:
https://github.com/asdf-format/asdf/pull/1508/commits/84f797ada168d9156e3b821a2a8fb24d9ec07a77:

File: asdf/asdf.py [hunk 2]
```diff
@@ -1344,7 +1344,57 @@ def write_to(
...
 
         """
-
+        with config_context() as config:
+            if "all_array_storage" in kwargs:
+                config.all_array_storage = kwargs["all_array_storage"]
+            if "all_array_compression" in kwargs:
+                config.all_array_compression = kwargs["all_array_compression"]
+            if "compression_kwargs" in kwargs:
+                config.all_array_compression_kwargs = kwargs["compression_kwargs"]
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/asdf.py:1352]: The tests will pass without these assignments but `_find_used_blocks` (which called `reserve_blocks`) won't be aware of the `all_array_*` settings without assigning them here (this is not important for any converters except for ndarray, which does not yet exist). These assignments might also not be needed if the block manager rework results in the removal of `reserve_blocks` (which I think should be one outcome). I'm happy to remove these assignments or leave them in as they're more of an attempt to prevent a possible future bug (which we're not testing for as we don't yet have an ndarray converter).

File: asdf/asdf.py [hunk 3]
```diff
@@ -1344,7 +1344,57 @@ def write_to(
...
+                uri=self._uri,
+                extensions=self.extensions,
+                version=self.version,
+                ignore_version_mismatch=self._ignore_version_mismatch,
+                ignore_unrecognized_tag=self._ignore_unrecognized_tag,
+                ignore_implicit_conversion=self._ignore_implicit_conversion,
+            )
+            naf._tree = self.tree  # avoid an extra validate
+            # copy over block storage and other settings
+            block_to_key_mapping = {v: k for k, v in self._blocks._data_to_block_mapping.items()}
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/asdf.py]: Good question.

Within this scope, the `block_to_key_mapping` is only used for blocks that don't have associated 'data' (and instead are relying on callbacks, like with the converters). This would mean that `_data_to_block_mapping` contains only key->block associations added by `assign_block_key`.

It does look possible for a Converter to call `assign_block_key` with the same index but different keys. This would be an error in how the custom converter was written as they currently cannot support sharing blocks across objects (I don't see how `find_block_index` could allow for writing the same block with multiple keys).

I separated the _data_to_block and _key_to_block mappings in this commit and added a test for the above case:
https://github.com/braingram/asdf/commit/389792df322791c29a05a60d74682736e6046bd2

File: asdf/asdf.py [hunk 4]
```diff
@@ -1344,7 +1344,57 @@ def write_to(
...
+            naf = AsdfFile(
+                {},
+                uri=self._uri,
+                extensions=self.extensions,
+                version=self.version,
+                ignore_version_mismatch=self._ignore_version_mismatch,
+                ignore_unrecognized_tag=self._ignore_unrecognized_tag,
+                ignore_implicit_conversion=self._ignore_implicit_conversion,
+            )
+            naf._tree = self.tree  # avoid an extra validate
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/asdf.py:1365]: That is a good point!

I added a test for this and modified `write_to` to deep copy the history and asdf_library sections prior to writing (and replace them after write_to is finished):
https://github.com/braingram/asdf/commit/8cedf33aaa4bf822e8b853712435d0facd7b8ee5

File: asdf/asdf.py [hunk 5]
```diff
@@ -1344,7 +1344,57 @@ def write_to(
...
+                    arr = b.data
+                    blk = naf._blocks[arr]
+                    blk._used = True
+                    naf.set_array_storage(arr, b.array_storage)
+                    naf.set_array_compression(arr, b.output_compression, **b.output_compression_kwargs)
+                else:  # this block does not have data
+                    key = block_to_key_mapping[b]
+                    blk = naf._blocks.find_or_create_block(key)
+                    blk._used = True
+                    blk._data_callback = b._data_callback
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/asdf.py:1394]: Thanks! I followed your suggestion adding a note to the docs about the 'repeatability' of the data callback.
https://github.com/braingram/asdf/commit/5c002455a0b03e83a8412040820af0ebb9fa816f

File: asdf/block.py [hunk 6]
```diff
@@ -80,7 +80,7 @@ def _add(self, block, key=None):
 
         if block._data is not None or key is not None:
             if key is None:
-                key = id(block._data)
+                key = id(util.get_array_base(block._data))
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/block.py]: Thanks for the catch. This was no longer needed (I believe it was a vestigial change from early work to update the blocks). I removed the call here:
https://github.com/braingram/asdf/commit/6597bede7dedec1f227f4000a99c35a89d5f54c9

File: asdf/block.py [hunk 7]
```diff
@@ -554,20 +555,31 @@ def get_external_uri(self, uri, index):
...
         reserved_blocks = set()
 
         for node in treeutil.iter_tree(tree):
-            # check that this object will not be handled by a converter
-            if not ctx.extension_manager.handles_type(type(node)):
+            if ctx.extension_manager.handles_type(type(node)):
+                converter = ctx.extension_manager.get_converter_for_type(type(node))
+                sctx = ctx._create_serialization_context()
+                tag = converter.select_tag(node, sctx)
+                for key in converter.reserve_blocks(node, tag, sctx):
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/block.py:572]: The original thought was that different tag versions might require different block usage. I guess one option might be to pass in the tag but not the serialization context but this felt at odds with the other parts of the Converter API. The serialization context would still need to be provided to `select_tag` so it's possible a converter could mess with blocks at that point as well.

One solution to this might be a more aware serialization context. Something that knows if this is a validation, a write, a read, etc and only allows modifications that are consistent with that operation (e.g. on validate, don't modify, allocate or free blocks). I made a brief attempt at something like this but ran into issues with the block manager (it's difficult to interact without producing a change in it's state).

So some options to address this might be:
- drop the serialization context (and possibly tag) argument to `reserve_blocks`
- leave in the argument(s) and add tests to cover their usage

As there is no current use of the serialization context in `reserve_blocks` I'm not sure what test should be written to cover it's usage. Are there examples of where this argument is used in other converters?

I'm good with either option at this point. I'm also not seeing `SerializationContext` listed in our user or developer api documentation but attempting to add a fix for that to this PR seems like scope creep.

File: asdf/block.py [hunk 8]
```diff
@@ -733,11 +751,8 @@ def find_or_create_block_for_array(self, arr):
         """
         from .tags.core import ndarray
 
-        if isinstance(arr, ndarray.NDArrayType) and arr.block is not None:
-            if arr.block in self.blocks:
-                return arr.block
-
-            arr._block = None
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/block.py:740]: Not really sure, it appears to have been added during a 'first pass at update support':
https://github.com/braingram/asdf/commit/edc6dfa7dd92bf2c64bf778e0788bcbce8b9f486
and currently appears to be the cause for this issue:
https://github.com/asdf-format/asdf/issues/1505

File: asdf/tags/core/integer.py [hunk 9]
```diff
@@ -76,7 +73,7 @@ def to_tree(cls, node, ctx):
 
             array = np.array(words, dtype=np.uint32)
             if node._storage == "internal":
-                cls._value_cache[ctx][abs_value] = array
+                cls._value_cache[abs_value] = array
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/tags/core/integer.py:76]: I'm good with removing it. I would first want to look to see if it's documented anywhere and if any of the compatibility or other tests expect this behavior.

It looks like the PR that updates Integer to a converter already drops the cache:
https://github.com/asdf-format/asdf/pull/1475

File: asdf/util.py [hunk 10]
```diff
@@ -529,3 +529,20 @@ class FileType(enum.Enum):
     ASDF = 1
     FITS = 2
     UNKNOWN = 3
+
+
+class BlockKey:
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/util.py:534]: No reason not to. I added an __eq__ method and test here:
https://github.com/braingram/asdf/commit/dd919be248e3eb772738a6679048508b223f2fbc

File: asdf/block.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/block.py]: Dropped the `ctx` arg here: 93c2eac16784bf490fee5e7851d94bdbc1509dd0

**Comment by braingram** [comment]
> I don't know what roman_datamodels is complaining about, I think this looks great

¯\_(ツ)_/¯

I think the roman failure was related to:
https://github.com/spacetelescope/roman_datamodels/pull/160

Which means it should go away with this most recent rebase.

I expect the devdeps jobs to fail because of:
https://github.com/asdf-format/asdf/pull/1532

**Comment by CagtayFabry** [comment]
If I understand correctly, this would solve https://github.com/asdf-format/asdf/issues/895 and possibly enable https://github.com/asdf-format/asdf/issues/1013 ?

I haven't looked too deep into it yet though

**Comment by braingram** [comment]
> If I understand correctly, this would solve #895 and possibly enable #1013 ?
> 
> I haven't looked too deep into it yet though

@CagtayFabry Thanks for taking a look!

I'm not sure if this PR solves #895 as the data callbacks for writing ASDF blocks are expected to return ndarrays (of dtype 'uint8'). However, the `BlockConverter` example in the tests does convert between ndarrays and bytes (using `frombuffer` and `tobytes`):
https://github.com/asdf-format/asdf/pull/1508/files#diff-8498843bd39a0657bc330de864a4ead415900e01f5bcdd15a17cb7af98bbee0fR25-R44

#1013 isn't addressed with this PR. There are some issues with the block management that make addressing this difficult (mainly that block settings, data and state are all stored in the same objects). I've been working on updating the block management in ASDF to allow us to move ndarray to a converter (and drop the legacy type system internally). I think that #1013 should be addressable with the rewrite as it more cleanly separates block options (like storage settings) from other aspects of block management. In addition to the internal changes, the `SerializationContext` will need to be updated to have methods similar to `get/set_array_storage/compression` etc.

**Comment by CagtayFabry** [comment]
I see, thank you for explaining @braingram

**Comment by braingram** [comment]
> I see, thank you for explaining @braingram

Certainly! To follow up, I added the following test to the branch with the block management rewrite:
https://github.com/braingram/asdf/blob/4bdaf70891cf61e025fd2ea93f56d75e8132eca7/asdf/_tests/_issues/test_1013.py
So far with the rewrite it is working (a few other things are still broken). Ignoring the private attribute access, does it look like that test covers #1013?

Also, any preference on the 'public' API? I'm inclined to have it match the methods in AsdfFile (get/set_array_storage/compression/compression_kwargs).

**Comment by braingram** [comment]
Newly failing weldx CI appears related to a new version of pint:
https://github.com/BAMWelDX/weldx/issues/872
